### PR TITLE
Modifying vm_metrics.py and vm_metrics_test.py to fetch ops metrics and Refactoring the code

### PR DESCRIPTION
**Pull request by etiksha**
Modifying vm_metrics.py to fetch ops latency, read bytes and error counts.
Adding Metric Class to generalize the code
Added tests in vm_metrics_test.py for the same.
Added test files.
Modifying build.sh to update gcsfuse version and adding experimental-stackdriver-export-interval flag

### CODE REVIEW

File: perfmetrics/scripts/vm_metrics/testdata/ops_mean_latency_response.json [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+          "bucket_counts": 19,
+          "bucket_counts": 18,
+          "bucket_counts": 17,
+          "bucket_counts": 302,
+          "bucket_counts": 760,
+          "bucket_counts": 1054,
+          "bucket_counts": 618,
+          "bucket_counts": 398,
+          "bucket_counts": 174,
+          "bucket_counts": 30,
```
Comment (vadlakondaswetha) [inline perfmetrics/scripts/vm_metrics/testdata/ops_mean_latency_response.json:433]: If its possible to add smaller data for tests, please do that.

File: perfmetrics/scripts/vm_metrics/vm_metrics.py [hunk 2]
```diff
@@ -1,92 +1,90 @@
...
-instance name and the mean period. Then it creates an instance of VmMetrics
-class and calls its methods with all the corresponding parameters, makes the api
-call, reads the response and returns a list of metric points where each point
-has the peak value and mean value of a mean period. Then it dumps all the data
-into a google sheet.
+"""Extracts required Google Cloud metrics using Monitoring V3 API call, parses
+   the API response into a list and writes to google sheet.
+
+   Takes VM instance name, interval start time, interval end time, alignment
+   period and fio test type as command line inputs.
```
Comment (vadlakondaswetha) [inline perfmetrics/scripts/vm_metrics/vm_metrics.py:5]: add a comment on  what are supported fio test types

File: perfmetrics/scripts/vm_metrics/vm_metrics.py [hunk 3]
```diff
@@ -1,92 +1,90 @@
...
   start_time_sec: int
   end_time_sec: int
 
-def _parse_metric_value_by_type(peak_value, peak_value_type) -> float:
-  if peak_value_type == 3:
-    return peak_value.double_value
-  elif peak_value_type == 2:
-    return peak_value.int64_value
+
+def _parse_metric_value_by_type(value, value_type) -> float:
```
Comment (vadlakondaswetha) [inline perfmetrics/scripts/vm_metrics/vm_metrics.py:51]: add a comment here pointing to value_type definition

File: perfmetrics/scripts/vm_metrics/vm_metrics.py [hunk 4]
```diff
@@ -100,133 +98,179 @@ def _validate_start_end_times(self, start_time_sec, end_time_sec):
...
+          'metric.type = "{metric_type}" AND metric.label.instance_name '
+          '={instance_name}'
+          ).format(metric_type=metric_type, instance_name=instance)
+
+    elif (metric_type[0:6] == 'custom'):
+      metric_filter = (
+          'metric.type = "{metric_type}" AND metric.labels.opencensus_task = '
+          'ends_with("{instance_name}")'
+          ).format(metric_type=metric_type, instance_name=instance)
+
```
Comment (vadlakondaswetha) [inline perfmetrics/scripts/vm_metrics/vm_metrics.py:142]: can you extract this get_metric_type into a seperate function and also add comment that what 0:7 / 0:6 mean?

File: perfmetrics/scripts/vm_metrics/vm_metrics.py [hunk 5]
```diff
@@ -100,133 +98,179 @@ def _validate_start_end_times(self, start_time_sec, end_time_sec):
...
+      test_type(str): The type of load test for which metrics are taken
     Returns: None
     """
     self._validate_start_end_times(start_time_sec, end_time_sec)
-    cpu_uti_data = self._get_metrics(start_time_sec, end_time_sec, instance,
-                                     period, CPU_UTI_METRIC, 1 / 100)
-    rec_bytes_data = self._get_metrics(start_time_sec, end_time_sec, instance,
-                                       period, RECEIVED_BYTES_COUNT_METRIC,
-                                       60000)
+    global TEST_TYPE
```
Comment (vadlakondaswetha) [inline perfmetrics/scripts/vm_metrics/vm_metrics.py:213]: you shouldn't declare global variables like this. 
test_type is anyways a param here, pass the same to required function and then call a method like this whereever you need this conversion. _get_ops_metric_method_name(test_type)

**Review by vadlakondaswetha — CHANGES_REQUESTED** [review]

File: perfmetrics/scripts/vm_metrics/vm_metrics.py (hunk 2 above)
Comment (etiksha) [reply perfmetrics/scripts/vm_metrics/vm_metrics.py:5]: done

File: perfmetrics/scripts/vm_metrics/vm_metrics.py (hunk 4 above)
Comment (etiksha) [reply perfmetrics/scripts/vm_metrics/vm_metrics.py:169]: done

File: perfmetrics/scripts/vm_metrics/vm_metrics.py (hunk 5 above)
Comment (etiksha) [reply perfmetrics/scripts/vm_metrics/vm_metrics.py]: done

File: perfmetrics/scripts/vm_metrics/vm_metrics.py [hunk 6]
```diff
@@ -1,92 +1,118 @@
...
+"""Extracts required Google Cloud metrics using Monitoring V3 API call, parses
+   the API response into a list and writes to google sheet.
+
+   Takes VM instance name, interval start time, interval end time, alignment
+   period, fio test type and google sheet name as command line inputs.
+   The supported fio test types are: read, write
+   Metrics extracted:
+   1.Peak Cpu Utilization(%)
+   2.Mean Cpu Utilization(%)
+   3.Peak Network Bandwidth(By/s)
```
Comment (vadlakondaswetha) [inline perfmetrics/scripts/vm_metrics/vm_metrics.py:10]: by is not the short form for bytes. Please dont use it anywhere. rename everywhere to bytes

File: perfmetrics/scripts/vm_metrics/vm_metrics.py (hunk 3 above)
Comment (vadlakondaswetha) [reply perfmetrics/scripts/vm_metrics/vm_metrics.py:58]: comment is still no added. i meant link to the code or doc where value_type is defined.

File: perfmetrics/scripts/vm_metrics/vm_metrics.py [hunk 7]
```diff
@@ -100,133 +126,176 @@ def _validate_start_end_times(self, start_time_sec, end_time_sec):
...
+      metric_filter = _get_metric_filter('custom', metric_type, instance)
 
+      # Adding extra filters:
+      if (metric_type == OPS_ERROR_COUNT_METRIC):
+        for fs_op in ERROR_COUNT_FILTERED_OUT_FS_OP:
+          metric_filter = ('{} AND metric.labels.fs_op != {}').format(metric_filter, fs_op)
+      elif (metric_type == OPS_LATENCY_METRIC):
+        fs_type = _get_ops_metric_method_name(test_type)
+        metric_filter = ('{} AND metric.labels.fs_op = {}').format(metric_filter, fs_type)
+    else:
```
Comment (vadlakondaswetha) [inline perfmetrics/scripts/vm_metrics/vm_metrics.py:175]: why not add this logic also in _get_metric_filter

**Review by vadlakondaswetha — APPROVED** [review]

**Review by vadlakondaswetha — APPROVED** [review]

**Pull request merged**